### ISSUE TITLE
[$8000] Compose box disappears when user edit message and press "New message" button.

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#15216]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Login with any account
2. Click into a chat that has many messages
3. Mark last message as unread
4. Scroll up so that the "unread" last message is out of view
5. Tap "Edit message" on any of your messages
6. Don't make any changes to the message
7. Tap on the "New Message" button at the top of the screen
8. You are brought down to the "unread" message. Note that the compose box is gone so you cannot write any new messages. 🔴 
9. Tap Back to return to chat list
10. Tap into the chat you are testing with again
11. Scroll up to the message you were editing
12. Note that it's still in edit mode and has the "Cancel" and "Save Changes" buttons showing 🔴

## Expected Result:
Result expected by Kavi
>Composer should be visible and scroll be should be on the right when you open a the report.

Result expected by Sonia
>If you tap "New Message" it should bring you to the unread message and should show you the composer at the bottom of the chat. When you leave the chat and go back the message should not remain in edit mode and you should see the composer at the bottom of the chat.

## Actual Result:
The Composer is still missing and when you tried to scroll, scroll bar is on the left.

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?

- [ ] iOS / native

**Version Number:** 1.2.72-0
**Reproducible in staging?:** y 
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/219454448-57745d6f-a326-423c-9a13-f2d81d565a89.MP4

https://user-images.githubusercontent.com/43996225/219454481-2f4eeb5d-334e-4d70-bcef-31bf67aab3e5.mov

**Expensify/Expensify Issue URL:**
**Issue reported by:** @hungvu193 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1676518791786289
 
[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0152a52b263a42cee8</li>
        <li>Upwork Job ID: 1629212143551795200</li>
        <li>Last Price Increase: 2023-05-15</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @sonialiap (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
@sonialiap Huh... This is 4 days overdue. Who can take care of this?

**Comment by sonialiap** [comment]
I cannot reproduce on iPhone 13 v 15.4


https://user-images.githubusercontent.com/17131195/220679545-2734d897-3c70-4d07-9d11-07d0162414bf.mp4

**Comment by sonialiap** [comment]
PrashantMangukiya was able to [reproduce](https://expensify.slack.com/archives/C049HHMV9SM/p1677083001534209?thread_ts=1676518791.786289&cid=C049HHMV9SM)

https://user-images.githubusercontent.com/17131195/220693389-7f69ee4a-1811-4110-9f10-1af77c9f9593.mp4

**Comment by sonialiap** [comment]
I can reproduce, thanks PrashantMangukiya for correcting the steps I was taking


https://user-images.githubusercontent.com/17131195/220880252-865dbfc4-c14a-4376-a62e-765bfedc6849.mp4

1. Click into a chat
2. Mark the bottom message as unread
3. Scroll up
4. Select "Edit" on a message you wrote
5. Don't make any changes to the message you're editing
6. Tap on the "New Message" button at the top of the screen
7. You are brought back down to the unread message ✅ . However, the compose box is gone so you cannot write any new messages 🔴 

Before taking the above steps
<img width="410" alt="image" src="https://user-images.githubusercontent.com/17131195/220880908-d04342e0-c016-4942-9423-63b7e53b95f4.png">

After taking the above steps
<img width="438" alt="image" src="https://user-images.githubusercontent.com/17131195/220882184-c42ba96f-c2f4-4876-8b1e-0be3a165622e.png">

**Comment by sonialiap** [comment]
@chiragsalian I agree with Kavi that the composer should be visible at the bottom of the chat even if you're in edit mode on a message higher up in the chat. I cannot reproduce the composer not showing up if you tap out and back into the chat, it shows up for me. I think that edit mode should be canceled when you tap out of the chat, but that is not the issue that was reported here.

**Comment by chiragsalian** [comment]
Interesting, yeah i can reproduce the issue. Marking it as external.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~0152a52b263a42cee8

**Comment by MelvinBot** [comment]
Current assignee @sonialiap is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`External`)

**Comment by MelvinBot** [comment]
Current assignee @chiragsalian is eligible for the External assigner, not assigning anyone new.

**Comment by PrashantMangukiya** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Whenever user edits message from top of the chat and press "New message" button, at that time it will scroll back to latest chat but chat compose box does not show i.e. chat compose box will disappear.

### What is the root cause of that problem?
Whenever we edit any message from the top of the chat at that time composer gets hidden. Meanwhile when we tap "New message" at that time we will scroll to the bottom of chat, but we do not take care to show composer by assuming that composer is already visible. So in this situation composer remain hidden.

https://github.com/Expensify/App/blob/94057baaf27d1f6c28277fc5cd713ee42ec10ada/src/pages/home/report/ReportActionsView.js#L297-L300

### What changes do you think we should make in order to solve the problem?

To solve this we have to call toggle report action compose view from `scrollToBottomAndMarkReportAsRead()` function  within `src/pages/home/report/ReportActionsView.js` file as shown under:

```javascript
 scrollToBottomAndMarkReportAsRead() { 
    toggleReportActionComposeView(true, this.props.isSmallScreenWidth);
    ....
 } 
```

So using this solution it will scroll down to new message and show compose box (instead of disappear) as shown in result video and also solve the scroll bar showing on the left issue.

With this solution it will not automatically take you out from edit mode. It is different issue and I think it should be handled separately. 

### What alternative solutions did you explore? (Optional)

None

<details>
<summary>Result iOS</summary>
<br/>

https://user-images.githubusercontent.com/7823358/219714017-50600451-a3df-4890-a7c6-88082e9a61f1.mp4
</details>

**Comment by MelvinBot** [comment]
📣 @PrashantMangukiya! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by allroundexperts** [comment]
I think it will make more sense to add `toggleReportActionComposeView` to the function [here](https://github.com/allroundexperts/Expensify/blob/151833d2956052a2ba6fe7e85042c31eba3b5c36/src/pages/home/report/ReportActionsView.js#L346). At this place, we can check if scroll offset is close to 0. If so, we can then show the composer input. This way, the composer bar will be visible even if the users scroll to the bottom of the page (without clicking that button). What do you think @PrashantMangukiya?

**Comment by PrashantMangukiya** [comment]
@allroundexperts  Thank you for suggestion. But adding content offset condition as shown below code is creating another weird situation. 

```javascript
trackScroll({nativeEvent}) {
        this.currentScrollOffset = -nativeEvent.contentOffset.y;
        if (this.currentScrollOffset > -300) {
            toggleReportActionComposeView(true, this.props.isSmallScreenWidth);
        }
        this.toggleFloatingMessageCounter();
    }
}
```

i.e. If user will go to edit record somewhere near the bottom, then it will also show composer and edit form both as shown in video. So I think checking the scroll offset may not be good idea. Thanks.

<details>
<summary>iOS Video</summary>

https://user-images.githubusercontent.com/7823358/221340200-5441edae-a5a3-4779-80e8-dd3105dc888a.mp4
</details>

**Comment by s77rt** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Main Composer not getting visible after clicking on "New messages" button while a Edit Composer is focused.

### What is the root cause of that problem?
On native, clicking outside a text input does not blur the text input.

_The reason this is hardly reproducible on Android is that Android automatically blurs the active input if it gets out of view._

### What changes do you think we should make in order to solve the problem?
Blur the text input on "New messages" button click. The easiest way of doing this is by dismissing the keyboard i.e. call `Keyboard.dismiss()` [here](https://github.com/Expensify/App/blob/9456e8c6640246498bae00a3fd4db3ab5228ea7e/src/pages/home/report/ReportActionsView.js#L297).

### What alternative solutions did you explore? (Optional)
None

**Comment by MelvinBot** [comment]
@chiragsalian @sonialiap @parasharrajat this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by MelvinBot** [comment]
@chiragsalian, @sonialiap, @parasharrajat Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MelvinBot** [comment]
@chiragsalian, @sonialiap, @parasharrajat Still overdue 6 days?! Let's take care of this!

**Comment by s77rt** [comment]
I don't see how this is related to this issue. This is actually the current behaviour on mWeb. We are just matching the behaviour here.

**Comment by PrashantMangukiya** [comment]
I think in this issue we have to handle composer box disappear problem when user go for edit message and press New message button to scroll the bottom of the list. Because here this problem occurs specific to iOS.

Because to change draft behaviour it needs separate discussion and change flow accordingly for all platforms. So change draft functionality can be handled separately.

Thanks @parasharrajat  and @s77rt for your time.

**Comment by MelvinBot** [comment]
@chiragsalian, @sonialiap, @parasharrajat Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by MelvinBot** [comment]
@chiragsalian, @sonialiap, @parasharrajat Eep! 4 days overdue now. Issues have feelings too...

**Comment by s77rt** [comment]
@parasharrajat Can you please clarify the cases? It seems to me unrelated to this issue that I'm thinking maybe my RCA is not clear enough. But in short whatever the behavior on mWeb is, it will be on native iOS. We are not doing anything here other than matching the behaviour since on mWeb clicking outside (on the button) will blur the input while on iOS native will not. We just blur that input manually.


| iOS - mWeb (Current) | iOS - Native (Before Fix) | iOS - Native (After Fix) |
|----------------------|---------------------------|--------------------------|
|    <video src="https://user-images.githubusercontent.com/16493223/225464219-7273471d-a01a-49b0-8fb1-dad937c9a179.mp4" />                  |        <video src="https://user-images.githubusercontent.com/16493223/225464339-5dbb7148-ae4b-4682-af9e-eb9bd93e4ada.mp4" />                   |      <video src="https://user-images.githubusercontent.com/16493223/225464352-262f463c-2b68-4cd4-ab89-1c3576f47ab6.mp4" />                    |


Unrelated: If you are wondering why on mWeb the Main Composer is focused after clicking the "New messages" button while on Native it's not; this is related to another factor; [willBlurTextInputOnTapOutside](https://github.com/Expensify/App/tree/366f0d064f50a614ca42aca3586beba6bfaba8c7/src/libs/willBlurTextInputOnTapOutside).

**Comment by MelvinBot** [comment]
@chiragsalian @sonialiap @parasharrajat this issue is now 4 weeks old and preventing us from maintaining WAQ, can you:
- Decide whether any proposals currently meet our guidelines and can be approved as-is today
- If no proposals meet that standard, please take this issue internal and treat it as one of your highest priorities
- If you have any questions, don't hesitate to start a discussion in #expensify-open-source

Thanks!

**Comment by MelvinBot** [comment]
Current assignee @parasharrajat is eligible for the Internal assigner, not assigning anyone new.

**Comment by sonialiap** [comment]
Just FYI, I'll be OOO Mar 17-25

**Comment by MelvinBot** [comment]
@chiragsalian, @sonialiap, @parasharrajat Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by chiragsalian** [comment]
So any update here @parasharrajat?

**Comment by PrashantMangukiya** [comment]
@parasharrajat Request to re-review my [proposal](https://github.com/Expensify/App/issues/15216#issuecomment-1444393118) above, in result video shown in my proposal, we can see that it will also dismiss keyboard smoothly, and scroll to bottom of the list. Thank you.

**Comment by PrashantMangukiya** [comment]
As per my suggested solution below is the code,
```javascript
  scrollToBottomAndMarkReportAsRead() {
      toggleReportActionComposeView(true, this.props.isSmallScreenWidth);
      ReportScrollManager.scrollToBottom();
      Report.readNewestAction(this.props.report.reportID);
  }
```
So once we call `toggleReportActionComposeView()` thereafter it call `ReportScrollManager.scrollToBottom();` so when list scroll to bottom it will hide keyboard after some threshold point (approx. half of the chat list). So this is the reason it hide keyboard when scroll down towards the bottom.

**Comment by PrashantMangukiya** [comment]
@parasharrajat  Yes I also realise that _Only keybord dismiss have some jitter_.  If we use both `toggleReportActionComposeView(true, this.props.isSmallScreenWidth);` and `Keyboard.dismiss()` it will do as expected for all i.e. Native iOS/Android and mWeb etc.  

Fingers crossed, let us see what internal team decide, please update me once decision come. Thanks for your time.

**Comment by s77rt** [comment]
Calling both `Keyboard.dismiss` and `toggleReportActionComposeView` may introduce a race condition where we focus the main composer twice and the second being an unwanted focus that may lead to unexpected results. I think we should only call `Keyboard.dismiss` and reconsider the use of `openReportActionComposeViewWhenClosingMessageEdit` [here](https://github.com/Expensify/App/blob/3f66c681cd1ad092507074799ba6400b58d1da85/src/pages/home/report/ReportActionItemMessageEdit.js#L287) and just use `toggleReportActionComposeView` as it was before.

`openReportActionComposeViewWhenClosingMessageEdit` has been introduced here by https://github.com/Expensify/App/pull/15201 to fix https://github.com/Expensify/App/issues/15086

**Comment by s77rt** [comment]
> Can you please share more details on this? Won't it break the fix https://github.com/Expensify/App/pull/15201.

What I meant is to fix the other issue in another way. Or to limit the functionality of "waiting for keyboard to hide" only for the case that was meant to be fixed on https://github.com/Expensify/App/issues/15086. May have to explore this some more.

> But I think it won't. toggleReportActionComposeView only changes the Onyx Prop value. So If we update the same Onyx prop twice, it won't affect it. The keyboard is already shown. React won't rerender if the prop is unchanged.

1. Click "New messages"
2. Main composer is focused (1st focus)
3. Focus something else (just use the app normally)
4. Suddenly main composer is focused again (2nd focus due to keyboard being hidden)

**Comment by s77rt** [comment]
I asked here https://github.com/Expensify/App/pull/15201#discussion_r1150858563 for a clarification

**Comment by chiragsalian** [comment]
> I will let CME handle it.

For now, i think 2k split 50-50 seems fair to me.

Thanks for staying on top of the issue @s77rt and @parasharrajat. I personally don't mind if you want to code out the proposal along with the changes suggested [here](https://github.com/Expensify/App/pull/15201/files#r1150858563). Just be careful and test the steps mentioned in this [PR](https://github.com/Expensify/App/pull/15201) too so that we aren't causing any regression.

Assigning @s77rt to work on the issue. If we should pause for any reason let me know.

**Comment by s77rt** [comment]
I think we should just dismiss the keyboard. The delay is caused by https://github.com/Expensify/App/pull/15201 technically this caused a (future?) regression. We will have to fix https://github.com/Expensify/App/issues/15086 in another way.

Would you agree if I raise a PR to just dismiss the keyboard and deal with the delay somewhere else.

**Comment by s77rt** [comment]
Reverting the PR of this issue won't fix the keyboard delay, there won't be a reason to revert the PR. There is no cycle here.

Nevertheless, let's fix both issues. @chiragsalian Please apply `Help Wanted` and `External` labels to get more proposals here.

**Comment by chiragsalian** [comment]
oh wtf, melvin removed the external. Confusing, thanks for the reminder. Switching it back.

**Comment by MelvinBot** [comment]
Current assignee @sonialiap is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignees @s77rt and @parasharrajat are eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @chiragsalian is eligible for the External assigner, not assigning anyone new.

**Comment by chiragsalian** [comment]
> to get more proposals here.

Are you not interested on working on this @s77rt? Asking because I assigned it to in assumption that you will begin working on a PR for this. Or do you prefer if I unassign you and we wait till we get more proposals?

**Comment by s77rt** [comment]
I'd be working on this but not explicitly. I'm interested in the proposals that we may get. The problem that I'm facing is that there is no info about the next focused element on native for blur event. So it's hard to conditionally open the main composer on native (currently we always open it, where on web there are conditions based on `event.relatedTarget`)

**Comment by Litande** [comment]
> IMO, I can perceive two issues from the reproduction steps.
> 
> 1. When you click on the `new messages` button, the composer is not available to write messages. As the edited message is deep inside the chat history, it is unclear why the composer is hidden. The user might not remember that he left a draft message on a report if he does not interact with that chat for a long time.
> 2. Second is that when we reopen the chat from the chat list, the previously edited messages are kept in edited mode. Again, after a long time, the user might not remember that there is a draft message for that report.
> 
> 2 is intentional to the point that we keep the comments in edited mode even though the report is closed. This is our draft comment feature. But I think a real problem could be that user will not know if there are any drafts for that report if a deeply nested message from history was edited in past.

@parasharrajat, @s77rt  hi considering 2 is intentional maibe the logic should be closing the editor if it gone from viewport, and mark the message with some kind of "draft" tag/word/sign if the message is different

**Comment by s77rt** [comment]
@Litande I not sure I the point of what you are suggesting. What we are looking for here is: If you click on "New messages" the chat should scroll to the bottom and the main composer should be visible immediately.

**Comment by Litande** [comment]
@s77rt
Sorry will try to explain a bit better.
in `ReportActionItemMessageEdit` on focus the following is called `toggleReportActionComposeView(false, this.props.isSmallScreenWidth);` which sets to main composer hidden when we are clicking New Messages nothing changes , this action isn't calling the blur effect on the current edited item.

i am saying considring the points mentioned by @parasharrajat the current behavior says:
* multiple instances of edit can exists if any of them selected hide main composer

i am suggesting why not make sure that only one textfield available those making it really easy to manage the main composer display ether composer is displayed or one of the items is edited, if you presed new messages, if scrolled away from the edited item, just close the state of editing mark it as contains draft and this will be a flag to main composer to show.

this suggestion may be not relevant due to some required logic, or because the task already assigned.

**Comment by Litande** [comment]
@parasharrajat i didn't mean to discard them, just to show regular ReportActionItemMessage with draft sign

but i believe such change is a bit too much for fixing the current issue.

@s77rt 
https://user-images.githubusercontent.com/13526647/228957017-7dddc7a1-3600-409b-b7f2-bf9e39297c00.mp4

**Comment by s77rt** [comment]
@Litande Are you cancelling the edit when clicking "New messages"? or When the input gets outside the view? Can we fix the bug but keep the input in the same state?

**Comment by s77rt** [comment]
Oh right, sorry about that. Unassigned myself.

**Comment by Litande** [comment]
@s77rt my approache to have only one edit state at a time, this exactly what controlls the main composer i don't believe it is good aproache to have 2+ inputs in the report view (including the main composer) in this example i have implmented a clear of edit on report on click - but i am sure i can achieve it if the edited item is out of view.

for it looks strange that we hide the main composer when editing an item and then don't hide the editing if we scroll to bottom and show the editor.

**Comment by s77rt** [comment]
@Litande Personally I don't think you answered my question. It's just foggy to me what you are trying to say here. I'm trying to understand how is the main composer getting visible again? `toggleReportActionComposeView` must be called somehow, my question is where we are calling it? I think you should follow the proposal template for this one. @parasharrajat is the C+ here, he will review your proposal.

**Comment by JediWattson** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Compose does not toggle on when "New Message" button is clicked while a message is being edited in iOS.

### What is the root cause of that problem?

It seems that when the button "New Message" is clicked there's no calls to toggle the composer for `toggleReportActionComposeView` this should be called in smaller screens since it's removed when a message is edited.

### What changes do you think we should make in order to solve the problem?


I would start with adding a prop that filters any drafts currently associated with that report ID.

```
withReportActionsDrafts({
        propName: 'draftKeys',
        transformValue: (drafts, props) => {
            const draftsReportKey = `${ONYXKEYS.COLLECTION.REPORT_ACTIONS_DRAFTS}${props.report.reportID}`;
            const draftKeys =  Object.keys(drafts).filter(key => key.includes(draftsReportKey));
            return draftKeys;
        },
    })
```

this will provide an array of strings known as `draftKeys` which can be used to essentially delete the draft when "New Messages" is pressed

```
    scrollToBottomAndMarkReportAsRead() {
        this.props.draftKeys.forEach(key => {
            const split = key.split("_");
            Report.saveReportActionDraft(split[1], split[2], '');    
        })

        toggleReportActionComposeView(true, this.props.isSmallScreenWidth);
        
        ReportScrollManager.scrollToBottom();
        ReportActionComposeFocusManager.focus();

        Report.readNewestAction(this.props.report.reportID);
    }
```

Here I'm splitting the draftKeys then calling Report.saveReportActionDraft with an empty string to remove the draft from editing mode. I then call toggleReportActionComposeView with isSmallScreenWidth which will fall through if this is false because the composer does not get removed in that case.

### What alternative solutions did you explore? (Optional)

I was thinking the discussion here is about the editor state. Should it use this.props.isSmallScreenWidth to only remove those drafts?

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by MelvinBot** [comment]
📣 @JediWattson! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by s77rt** [comment]
I'm not totally against the headed approach but I don't think it's ideal:
- Not generic: The solution is only for Edit Composer, it's hard to extend the functionality to other inputs e.g. `saveButtonID`, `cancelButtonID` and `emojiButtonID` will still show the Main composer on Native (but not on Web).
- Does not work for programmatically focused inputs (`onPressIn` and `onPressOut` won't be called).

I think what we should do here is just to keep the Main Composer always visible i.e. remove the whole `toggleReportActionComposeView` functionality, if we are not doing something right, it's better not to do it at all.

Or better yet, refactor the edit functionality, I have asked in Slack https://expensify.slack.com/archives/C01GTK53T8Q/p1680398723148189

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by s77rt** [comment]
I think we still need more :eyes: on this. I have just raised the concern of removing `toggleReportActionComposeView`.

**Comment by mdneyazahmad** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When editing a message, if we click "New message" button, it does not open the main composer.

### What is the root cause of that problem?

In order to show the main composer, it has to somehow call `toggleReportActionComposeView`. In the current implementation this is handled in `onBlur` of edit text composer component. When "New message" button is pressed, this does not blur the edit text composer. In turn the code in the blur function is not executed, and this issue appears. This issue is not exclusive with "New message" button. If there is any button, it will have same behaviour.

There are some inconsistencies between iOS and Android in the way it handles the scroll of a `FlatList`. If there is a `TextInput` inside a `FlatList` item. It loses focus on android as soon as it goes out of visible area, thus blur event is fired, and this is not the case with iOS. `TextInput` does not lose focus event if goes out of visible area, thus `onBlur` is not called.

This issue also exists on android, but the reproduction steps is little different, as stated the edited text input should remain inside the visible area to reproduce this issue on android. For that do not scroll deep inside the message list. 


https://user-images.githubusercontent.com/77761491/231119191-2e5580ec-3cc4-46f4-ac19-f4395d04282c.mp4


https://user-images.githubusercontent.com/77761491/231119028-5d89c6cc-595e-4025-9cba-ad862e1d0370.mp4


### What changes do you think we should make in order to solve the problem?


1. Dismiss keyboard when "New message" button is pressed, this will execute the required code (`onBlur`) to show the composer.

Following this creates an issue as stated in the previous proposal.

To fix that:
1. Remove `openReportActionComposeViewWhenClosingMessageEdit`

https://github.com/Expensify/App/blob/d196a1217754ffab2e10ba28ba3c88f0b6391bf5/src/pages/home/report/ReportActionItemMessageEdit.js#L292

2. Add this condition if there is next edit text input focused, return or else execute `toggleReportActionComposeView`

```javascript
                            if (this.messageEditInput === relatedTargetId) {
                                return;
                            }

                            //  get currently focused input, returns null if no input is focused
                            const el = TextInput.State.currentlyFocusedInput();

                            // if this is focused verify that is an instance of `ReportActionItemMessageEdit`
                            if (el && el.isFocused() && el.getName() === 'ReportActionItemMessageEdit') {
                                return;
                            }

                            toggleReportActionComposeView(true, this.props.isSmallScreenWidth);
```

```javascript
                    <Composer
                        multiline
                        ref={(el) => {
                            // here add the `getName` props to verify this is exactly this reference (also works without this, but better to add this check)
                            _.assign(el, {
                                getName: () => 'ReportActionItemMessageEdit',
                            });
                            this.textInput = el;
                            this.props.forwardedRef(el);
                        }}
```

### What alternative solutions did you explore? (Optional)

### Result

https://user-images.githubusercontent.com/77761491/231161701-2c6e370b-db43-4c87-aba5-7b89b72fd5ff.mp4

**Comment by s77rt** [comment]
I think @mdneyazahmad proposal is solid from programmatic focus perspective but it's still not generic, `TextInput.State.currentlyFocusedInput` only work for text inputs https://github.com/facebook/react-native/blob/6eebc997426863cf2e24aee1f72d48d4a314b67d/packages/react-native/Libraries/Components/TextInput/TextInput.d.ts#L822-L826.

If only we had an equivalent for `document.activeElement` for Native.

**Comment by MelvinBot** [comment]
@chiragsalian, @sonialiap, @parasharrajat Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by chiragsalian** [comment]
@parasharrajat, friendly bump, any update here?

**Comment by MelvinBot** [comment]
@chiragsalian, @sonialiap, @parasharrajat Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by sonialiap** [comment]
No update in the [slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1680398723148189)

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by sonialiap** [comment]
@parasharrajat do we have an update for this issue?

**Comment by michaelhaxhiu** [comment]
Seems like @s77rt opened a GH https://github.com/react-native-community/discussions-and-proposals/issues/640 and @parasharrajat needs to review. Can we aim to do that in the next 1-2 days?

**Comment by sonialiap** [comment]
Not overdue

**Comment by sonialiap** [comment]
@parasharrajat seems like we're waiting on your review for this issue, any updates?

**Comment by sonialiap** [comment]
Sounds good, thanks for the update

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by sonialiap** [comment]
I cannot reproduce anymore either. Will check in slack if the reporter can reproduce this. It's possible it was fixed by something else along the way.
